### ISSUE TITLE
How to manage memory leak issues in Sinks Flux

### ISSUE THREAD

**Issue by ghkdqhrbals** [ghkdqhrbals/spring-chatting-server#59]
## **Description**

Since we using Reactive Stream, we should explicitly set and manage termination conditions. This is very considerable, time-consuming tasks. If we miss some terminations of stream, memory leakage will be occured. I will explain the current process of using reactive stream.

we use Sinks.Many HashMap to communicate with other threads.

> * `sinkMap` saved in Methods Area which is not the target of GC.
> * The objects in `sinkMap` are saved in Heap Area which is the target of GC.

1. We set `sinkMap` as ConcurrentHashMap for thread safe.

```java
@Configuration
@Slf4j
public class AsyncConfig  {
    public static ConcurrentHashMap<String, Sinks.Many<Object>> sinkMap = new ConcurrentHashMap<>();
    ...
}
```

2. We add `Sinks.many().multicast().onBackpressureBuffer()` in `sinkMap` when the server get request from the client.

```java
@PostMapping(value = "/user", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<?> addUser(@RequestBody RequestUser req) throws InterruptedException {
        ...
        AsyncConfig.sinkMap.put(req.getUserId(), Sinks.many().multicast().onBackpressureBuffer());
        ...
}
```

3. We take this `backpressureBuffer` out in other threads. Using `.tryEmitNext()`, other threads can send values to Reactive Stream.

```java
    // kafka listener thread
    private void checkCurrentStatus(UserTransactions userTransactions,String userId) {
        if (!userTransactions.getChatStatus().equals(UserResponseStatus.USER_APPEND.name())
                && !userTransactions.getCustomerStatus().equals(UserResponseStatus.USER_APPEND.name())
                && !userTransactions.getUserStatus().equals(UserStatus.USER_INSERT.name())
                && !userTransactions.getUserStatus().equals(UserStatus.USER_DELETE.name())){
            AsyncConfig.sinkMap.get(userId).tryEmitNext(userTransactions);
            AsyncConfig.sinkMap.get(userId).tryEmitComplete();
            AsyncConfig.sinkMap.remove(userId);
        }
        else{
            AsyncConfig.sinkMap.get(userId).tryEmitNext(userTransactions);
        }
    }
```

4. After all threads finish to send values, `.tryEmitComplete` or `. tryEmitError()` is called.

```java
@KafkaListener(topics = KafkaTopic.userRes, containerFactory = "userKafkaListenerContainerFactory", concurrency = KafkaTopicPartition.userRes)
    public void listenUser(UserResponseEvent req) {
        userService.updateStatus(req).exceptionally(e->{
            log.info("There is no event transaction");
            AsyncConfig.sinkMap.get(req.getUserId()).tryEmitError(e); // 
            AsyncConfig.sinkMap.remove(req.getUserId()); // removing object in hashMap
            return null;
        });
    }
```

5. And the Reactive Stream that is created by `Sinks.many()` will be deconstructed.

6. Now the resource of Reactive Stream is returned, and remove the object located in `sinkMap`.

**Since reactive stream don't have automated termination, we need to keep our eyes on track every process.**
> But Flux has automated termination like `.take()`, `.timeout(Duration.ofSeconds(1))`

**Comment by ghkdqhrbals** [comment]
### Details about the `Sink`

The `Sink` created using `Sinks.many().multicast()` allows multicasting of data to multiple subscribers
> In detail, `Sinks.many()` return special Sink that **can produce multiple stream values**, and by adding `.multicast()`, **multiple subscriber can connect** to that Sink.
This multicasting capability ensures that multiple subscribers can receive the same data stream simultaneously.

`Backpressure` is an important mechanism in controlling the flow of data between such a Sink and its subscribers, especially when the processing speeds of the subscribers differ.

If a particular subscriber cannot process data quickly enough, the `backpressure` mechanism allows **controlling the data delivery rate** for that subscriber. `onBackpressureBuffer()` is one of the specific methods to handle backpressure situations. It uses an internal buffer to temporarily store data. If necessary, it can either **evict older data** or **refuse new data** that exceeds the buffer size.

In conclusion, `Sinks.many().multicast()` serves to relay data to multiple subscribers, and `onBackpressureBuffer()` manages the backpressure of the data flow.

**Comment by ghkdqhrbals** [comment]
Code changes in #63
